client-runtime/src/main/java/com/microsoft/rest/v3/RestProxy.java

```diff
@@ -406,93 +364,61 @@ private static Exception instantiateUnexpectedException(Class<? extends RestExce
-                                } catch (IllegalAccessException iae) {
-                                    throw reactor.core.Exceptions.propagate(iae);
-                                } catch (InvocationTargetException ite) {
-                                    throw reactor.core.Exceptions.propagate(ite);
-                                } catch (InstantiationException ie) {
-                                    throw reactor.core.Exceptions.propagate(ie);
-                                }
-                            })
-                            .switchIfEmpty(Mono.defer((Supplier<Mono<RestResponseBase<?, ?>>>) () -> {
-                                try {
-                                return Mono.just(responseConstructor.newInstance(response.sourceResponse().request(), responseStatusCode, deserializedHeaders, rawHeaders, null));
-                                } catch (IllegalAccessException iae) {
-                                    throw reactor.core.Exceptions.propagate(iae);
-                                } catch (InvocationTargetException ite) {
-                                    throw reactor.core.Exceptions.propagate(ite);
-                                } catch (InstantiationException ie) {
-                                    throw reactor.core.Exceptions.propagate(ie);
-                                }
-                            }));
-                }
-            } else {
-                // For now we're just throwing if the Maybe didn't emit a value.
-                asyncResult = handleBodyReturnType(response, methodParser, entityType);
-            }
+    private RestResponse<?> createResponse(HttpDecodedResponse response, Type entityType, Object bodyAsObject) {
+        final HttpResponse httpResponse = response.sourceResponse();
+        final HttpRequest httpRequest = httpResponse.request();
+        final int responseStatusCode = httpResponse.statusCode();
+        final HttpHeaders responseHeaders = httpResponse.headers();
 
-            return asyncResult;
-        } catch (ReflectiveOperationException e) {
-            throw new RuntimeException(e);
+        // determine the type of response class. If the type is the 'RestResponse' interface, we will use the
+        // 'RestResponseBase' class instead.
+        Class<? extends RestResponse<?>> cls = (Class<? extends RestResponse<?>>) TypeUtil.getRawClass(entityType);
+        if (cls.equals(RestResponse.class)) {
+            cls = (Class<? extends RestResponse<?>>)(Object) RestResponseBase.class;
+        }
+
+        // we assume there is a single constructor for each type, and just take the first
+        Constructor<? extends RestResponse<?>> ctor = (Constructor<? extends RestResponse<?>>) cls.getConstructors()[0];
+
+        // create the constructor args array
+        Object[] args = new Object[ctor.getParameterCount()];
+        args[0] = httpRequest;
+        args[1] = responseStatusCode;
+        args[2] = responseHeaders;
```

Comment:
Maybe add a check that the `ctor.getParameterCount()` be at least 3, and throw an exception if not or a Debug.assert (i don't know if those exist in Java).

Reply:
SDK implementors or others using our RestProxy can extends from one of the runtime/common response types and specify the extended type as the return type of their proxy method which we instantiate using reflection. In this case it would be good to have a check for -

1. At least one constructor and
2. with at least 3 parameters and
3. Even more defensive check for exact types of first 3 params (HttpRequest.class, Integer.TYPE, HttpHeaders.class)

Reply:
Good ideas - I'll improve this today.